client/state/data-layer/extensions-middleware.js

```diff
@@ -1,101 +1,58 @@
+
+	if ( ! omitted ) {
+		debug( `Trying to remove nonexistent handlers for ${ extensionName }` );
+	}
+
+	// Set the module globals for convenience
+	extensionHandlers = remainingHandlers;
+	handlerMiddleware = buildMiddleware( extensionHandlers );
+
+	return extensionHandlers;
+}
+
+export function buildMiddleware( handlersByExtension ) {
+	const allHandlers = Object.values( handlersByExtension ).reduce(
+		( accumulated, handlers ) => {
+			return mergeHandlers( accumulated, handlers );
+		},
+		{}
 	);
 
-	return [ key, newKeyHandlers ];
-};
-
-const stripHandlers = ( handlers, handlersToStrip ) => {
-	return fromPairs(
-		toPairs( handlers ).map(
-			( [ key, keyHandlers ] ) => stripHandlersForKey( key, keyHandlers, handlersToStrip[ key ] )
-		)
-	);
-};
-
-export const extensionsHandlers = () => {
-	let handlers = {};
-
-	/**
-	 * Dynamically add handlers to this middleware.
-	 *
-	 * @function
-	 * @param {Object} handlersToAdd The actions to be added.
-	 * @return {Function} removeHandlers, removes the handlers that were added.
-	 */
-	const addHandlers = ( handlersToAdd ) => {
-		handlers = mergeHandlers( handlers, handlersToAdd );
-
-		return function removeHandlers() {
-			handlers = stripHandlers( handlers, handlersToAdd );
-		};
-	};
+	return middleware( allHandlers );
+}
 
-	/**
-	 * WPCOM Extensions Middleware
-	 *
-	 * Intercepts actions requesting data for extensions
-	 * and passes them to the appropriate handler.
-	 *
-	 * This code patterns off of datalayer wpcom-api-middleware
-	 * @see state/data-layer/wpcom-api-middleware
-	 *
-	 * @param {Object} store Common `store => next => action` pattern for middlewares.
-	 * @return {Function} middleware handler
-	 */
-	const middleware = store => next => {
-		const localNext = action => next( local( action ) );
-
-		/**
-		 * Middleware handler
-		 *
-		 * @function
-		 * @param {Object} action Redux action
-		 * @returns {undefined} please do not use
-		 */
-		return action => {
-			const handlerChain = handlers[ action.type ];
-
-			// if no handler is defined for the action type
-			// then pass it along the chain untouched
-			if ( ! handlerChain ) {
-				return next( action );
-			}
-
-			const meta = action.meta;
-			if ( meta ) {
-				const dataLayer = meta.dataLayer;
-
-				// if the action indicates that we should
-				// bypass the data layer, then pass it
-				// along the chain untouched
-				if ( dataLayer && true === dataLayer.doBypass ) {
-					return next( action );
-				}
-			}
-
-			return handlerChain.forEach( handler => handler( store, action, localNext ) );
-		};
-	};
-
-	return { middleware, addHandlers };
+/**
+ * Extensions Middleware
+ * @function
+ * @param {Object} store The store for the middleware chain.
+ * @returns {Function} The next => action part of the middleware.
+ */
+export default store => next => action => {
+	return handlerMiddleware( store )( next )( action );
 };
```

Comment:
do you find it at all odd to use currying arrow-function syntax for all but the last one?

```js
export default store => next => action => handlerMiddleware( store )( next )( action );
```